floris/tools/wind_data.py

```diff
@@ -0,0 +1,529 @@
+        freq_table = wind_rose_resample.freq_table
+
+        # Set up figure
+        if ax is None:
+            _, ax = plt.subplots(subplot_kw={"polar": True})
+
+        # Get a color array
+        color_array = cm.get_cmap(color_map, len(ws_bins))
+
+        for wd_idx, wd in enumerate(wd_bins):
+            rects = []
+            freq_table_sub = freq_table[wd_idx, :].flatten()
+            for ws_idx, ws in reversed(list(enumerate(ws_bins))):
+                plot_val = freq_table_sub[:ws_idx].sum()
+                rects.append(
+                    ax.bar(
+                        np.radians(wd),
+                        plot_val,
+                        width=0.9 * np.radians(wd_step),
+                        color=color_array(ws_idx),
+                        edgecolor="k",
+                    )
+                )
+
+        # Configure the plot
+        ax.legend(reversed(rects), ws_bins, **legend_kwargs)
+        ax.set_theta_direction(-1)
+        ax.set_theta_offset(np.pi / 2.0)
+        ax.set_theta_zero_location("N")
+        ax.set_xticks(np.arange(0, 2 * np.pi, np.pi / 4))
+        ax.set_xticklabels(["N", "NE", "E", "SE", "S", "SW", "W", "NW"])
+
+        return ax
+
+
+class TimeSeries(WindDataBase):
+    """
+    In FLORIS v4, the TimeSeries class is used to drive FLORIS and optimization
+    operations in which the inflow is by a sequence of wind direction, wind speed
+    and turbulence intensitity values
+
+    Args:
+        wind_directions: NumPy array of wind directions (NDArrayFloat).
+        wind_speeds: NumPy array of wind speeds (NDArrayFloat).
+        turbulence_intensity:  NumPy array of wind speeds (NDArrayFloat, optional).
+            Defatuls to None
+        prices:  NumPy array of electricity prices (NDArrayFloat, optional).
+            Defatuls to None
+
+    """
+
+    def __init__(
+        self,
+        wind_directions,
+        wind_speeds,
+        turbulence_intensity=None,
+        prices=None,
+    ):
+        # Wind speeds and wind directions must be the same length
+        if len(wind_directions) != len(wind_speeds):
+            raise ValueError("wind_directions and wind_speeds must be the same length")
+
+        self.wind_directions = wind_directions
+        self.wind_speeds = wind_speeds
+        self.turbulence_intensity = turbulence_intensity
+        self.prices = prices
+
+        # Record findex
+        self.n_findex = len(self.wind_directions)
+
+    def unpack(self):
+        """
+        Unpack the time series data in a manner consistent with wind rose unpack
+        """
+
+        # to match wind_rose, make a uniform frequency
+        uniform_frequency = np.ones_like(self.wind_directions)
+        uniform_frequency = uniform_frequency / uniform_frequency.sum()
+
+        return (
+            self.wind_directions.copy(),
+            self.wind_speeds.copy(),
+            uniform_frequency,
+            self.turbulence_intensity,  # can be none so can't copy
```

Comment:
What is the impact of copying the other arrays but not these? in other words, is it safe to _not_ copy these?

Reply:
@paulf81 and I spoke a bit about this, and decided it was ok _not_ to copy `wind_directions` and `wind_speeds`, as `calculate_wake()` is not expected to modify these in any way (so it's ok to point to the same memory)

Reply:
Agreed!